internal/controller/kubecertagent/deployer.go

```diff
@@ -0,0 +1,273 @@
+	// the CredentialIssuer.
+	if newestControllerManager == nil {
+		c.log.Info("could not find a healthy kube-controller pod", "candidates", len(controllerManagerPods))
+		return issuerconfig.UpdateStrategy(
+			ctx.Context,
+			c.credentialIssuerLocationConfig.Name,
+			c.credentialIssuerLabels,
+			c.pinnipedAPIClient,
+			strategyError(c.clock, constable.Error("could not find a healthy kube-controller-manager pod")),
+		)
+	}
+
+	// Build the expected Deployment based on the kube-controller-manager Pod as a template.
+	c.log.Info("building expected Deployment based on kube-controller-manager Pod", "pod", klog.KObj(newestControllerManager))
+	expectedDeployment := newAgentDeployment(c.agentPodConfig, newestControllerManager)
+
+	// Try to get the existing Deployment, if it exists.
+	existingDeployment, err := c.agentDeploymentInformer.Lister().Deployments(expectedDeployment.Namespace).Get(expectedDeployment.Name)
+	notFound := k8serrors.IsNotFound(err)
+	if err != nil && !notFound {
+		return fmt.Errorf("informer cannot get agent deployments: %w", err)
+	}
+
+	// If the Deployment did not exist, create it and be done.
+	if notFound {
+		c.log.Info("creating new Deployment", "deployment", klog.KObj(expectedDeployment))
+		_, err := c.k8sClient.AppsV1().Deployments(expectedDeployment.Namespace).Create(ctx.Context, expectedDeployment, metav1.CreateOptions{})
+		return err
+	}
+
+	// If the existing Deployment already matches our desired spec, we're done.
+	if apiequality.Semantic.DeepDerivative(expectedDeployment, existingDeployment) {
+		c.log.Info("existing Deployment is already up to date", "deployment", klog.KObj(expectedDeployment))
+		return nil
+	}
+
+	// Otherwise update the spec of the Deployment to match our desired state and update it.
+	c.log.Info("updating existing Deployment", "deployment", klog.KObj(expectedDeployment))
+	expectedDeployment.Spec.DeepCopyInto(&existingDeployment.Spec)
+	existingDeployment.ObjectMeta = mergeObjectMeta(existingDeployment.ObjectMeta, expectedDeployment.ObjectMeta)
+	_, err = c.k8sClient.AppsV1().Deployments(expectedDeployment.Namespace).Update(ctx.Context, existingDeployment, metav1.UpdateOptions{})
+	return err
+}
+
+// newestRunningPod takes a list of pods and returns the newest one with status.phase == "Running".
+func newestRunningPod(pods []*corev1.Pod) *corev1.Pod {
+	// Compare two pods based on creation timestamp, breaking ties by name
+	newer := func(a, b *corev1.Pod) bool {
+		if a.CreationTimestamp.Time.Equal(b.CreationTimestamp.Time) {
+			return a.Name < b.Name
+		}
+		return a.CreationTimestamp.After(b.CreationTimestamp.Time)
+	}
+
+	var result *corev1.Pod
+	for _, pod := range pods {
+		if pod.Status.Phase == corev1.PodRunning && (result == nil || newer(pod, result)) {
+			result = pod
+		}
+	}
+	return result
+}
+
+func int32ptr(i int32) *int32 { return &i }
+func int64ptr(i int64) *int64 { return &i }
```

Comment:
Kube util code should have helpers for all of these.

Reply:
Ah yeah, in `k8s.io/utils/pointer`. I thought I remembered them existing but it was so easy to just rewrite them. Fixed.